Frameworks/Foundation/_NSInvocation.x86.mm

```diff
@@ -0,0 +1,207 @@
+#import <objc/encoding.h>
+
+enum return_type {
+    RETURN_TYPE_NONE,
+    RETURN_TYPE_SINT8,
+    RETURN_TYPE_UINT8,
+    RETURN_TYPE_SINT16,
+    RETURN_TYPE_UINT16,
+    RETURN_TYPE_INT32, // Only one: no need to sign-extend
+    RETURN_TYPE_INT64, // Only one: no need to sign-extend
+    RETURN_TYPE_POINTER,
+    RETURN_TYPE_FLOAT,
+    RETURN_TYPE_DOUBLE,
+    RETURN_TYPE_STRUCT,
+};
+
+struct x86Frame {
+    void* ebp;
+    void* retp;
+    void* ebx;
+    unsigned int rettype;
+    size_t retlen;
+    void* retvalp;
+    void* funp;
+};
+
+static return_type _getReturnType(const char* typeEncoding) {
+    switch (typeEncoding[0]) {
+        case _C_CHR:
+            return RETURN_TYPE_SINT8;
+        case _C_SHT:
+            return RETURN_TYPE_SINT16;
+        case _C_UCHR:
+            return RETURN_TYPE_UINT8;
+        case _C_USHT:
+            return RETURN_TYPE_UINT16;
+        case _C_BOOL:
+        // On Win32, due to Windows.h, BOOL is 4 bytes wide.
+        // Fall through.
+        case _C_INT:
+        case _C_LNG:
+        case _C_UINT:
+        case _C_ULNG:
+            return RETURN_TYPE_INT32;
+        case _C_LNG_LNG:
+        case _C_ULNG_LNG:
+            return RETURN_TYPE_INT64;
+        case _C_ID:
+        case _C_CLASS:
+        case _C_PTR:
+        case _C_SEL:
+        case _C_CHARPTR:
+            return RETURN_TYPE_POINTER;
+        case _C_STRUCT_B: {
+            size_t size = objc_sizeof_type(typeEncoding);
+
+            switch (size) {
+                case 8:
+                    return RETURN_TYPE_INT64;
+                case 4:
+                    return RETURN_TYPE_INT32;
+                case 2:
+                    return RETURN_TYPE_UINT16;
+                case 1:
+                    return RETURN_TYPE_UINT8;
+                default:
+                    return RETURN_TYPE_STRUCT;
+            }
+        }
+        case _C_FLT:
+            return RETURN_TYPE_FLOAT;
+        case _C_DBL:
+            return RETURN_TYPE_DOUBLE;
+        case _C_VOID:
+            return RETURN_TYPE_NONE;
+    }
+    return RETURN_TYPE_NONE;
+}
+
+_NSInvocationCallFrame::_NSInvocationCallFrame(NSMethodSignature* methodSignature)
+    : _methodSignature(methodSignature), _offset(0), _structReturn(false), _allocationExtents([_methodSignature numberOfArguments]) {
+    _returnType = _getReturnType([_methodSignature methodReturnType]);
+    _returnLength = [_methodSignature methodReturnLength];
+
+    // 1, 2, 4, and 8-byte structs are returned in registers.
+    if (_returnType == RETURN_TYPE_STRUCT) {
+        _structReturn = true;
+        _structReturnExtent = _allocateArgument("^v");
+    } else if (_returnType != RETURN_TYPE_NONE) {
+        // Promote all non-stret return lengths to one machine word.
+        _returnLength = std::max(sizeof(uintptr_t), _returnLength);
+    }
+
+    unsigned int nArguments = [_methodSignature numberOfArguments];
+    for (unsigned int i = 0; i < nArguments; ++i) {
```

Comment:
(in a couple spots)

---
In reply to: [77061000](https://github.com/Microsoft/WinObjC/pull/903#discussion_r77061000) [](ancestors = 77061000)